web/src/auth/CasLogout.js

```diff
@@ -35,24 +35,53 @@ class CasLogout extends React.Component {
   UNSAFE_componentWillMount() {
     const params = new URLSearchParams(this.props.location.search);
 
+    const logoutInterval = 100;
+    const logoutTimeOut = (s, redirectUri, linkSoftFlag) => {
+      setTimeout(() => {
+        AuthBackend.getAccount().then((accountRes) => {
+          if (accountRes.status === "ok") {
+            AuthBackend.logout().then((logoutRes) => {
+              if (logoutRes.status === "ok") {
+                const redirectData = logoutRes.data2;
+                if (redirectData !== null && redirectData !== undefined && redirectData !== "") {
+                  logoutTimeOut(s, redirectData, false);
+                } else if (params.has("service")) {
+                  logoutTimeOut(s, params.get("service"), false);
+                } else {
+                  logoutTimeOut(s, `/cas/${this.state.owner}/${this.state.applicationName}/login`, true);
+                }
+              } else {
+                Setting.showMessage("error", `Failed to log out: ${logoutRes.msg}`);
+              }
+            });
+          } else {
+            if (linkSoftFlag) {
+              Setting.goToLinkSoft(this, redirectUri);
+            } else {
+              Setting.goToLink(redirectUri);
+            }
+            Setting.showMessage("success", "Logged out successfully");
+            this.props.onUpdateAccount(null);
+          }
+        });
+      }, s);
```

Comment:
```js
const logoutTimeOut = (redirectUri) => {
      setTimeout(() => {
        AuthBackend.getAccount().then((accountRes) => {
          if (accountRes.status === "ok") {
            AuthBackend.logout().then((logoutRes) => {
              if (logoutRes.status === "ok") {
                const redirectData = logoutRes.data2;
                logoutTimeOut(redirectData);
              } else {
                Setting.showMessage("error", `Failed to log out: ${logoutRes.msg}`);
              }
            });
          } else {
            Setting.showMessage("success", "Logged out successfully");
            this.props.onUpdateAccount(null);
            if (redirectUri !== null && redirectUri !== undefined && redirectUri !== "") {
              Setting.goToLink(redirectUri);
            } else if (params.has("service")) {
              Setting.goToLink(params.get("service"));
            } else {
              Setting.goToLinkSoft(this, `/cas/${this.state.owner}/${this.state.applicationName}/login`);
            }
          }
        });
      },logoutInterval);
```
I think this might be easier to understand.

Reply:
I updated the timeout implementation. Could you please review the code again?